File: proxied-record-forces-ttl/python/proxied_record_forces_ttl.py
"""Detect a proxied Cloudflare record whose intended config expects a
custom TTL, which Cloudflare will always silently coerce to 1
(Automatic), and optionally repair the zone by either accepting
ttl: 1 in policy, or unproxying the record to keep a custom TTL.

Safe by default. Set DRY_RUN=false to let it write.

Env vars:
  DNS_DOMAIN               the domain to check, e.g. "example.com"
  CLOUDFLARE_API_TOKEN     Cloudflare API token (only needed for repair)
  CLOUDFLARE_ZONE_ID       Cloudflare zone id (only needed for repair)
  DRY_RUN                  default "true"; set to "false" to actually write
  REPAIR_POLICY            "accept_automatic" or "unproxy", default "accept_automatic"
"""
import os
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("proxied_record_forces_ttl")

DNS_DOMAIN = os.environ.get("DNS_DOMAIN", "example.com")
CLOUDFLARE_API_TOKEN = os.environ.get("CLOUDFLARE_API_TOKEN", "")
CLOUDFLARE_ZONE_ID = os.environ.get("CLOUDFLARE_ZONE_ID", "")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

CF_API = "https://api.cloudflare.com/client/v4"


def diagnose_ttl_proxy_mismatch(intended, live):
    """Pure decision function. No I/O.

    intended, live: {"ttl": int, "proxied": bool}

    Returns a mismatch reason string, or None if consistent.
      - If live.proxied is True and live.ttl != 1: impossible/stale-cache state.
      - If intended.proxied is True and intended.ttl not in (1, None):
        config is invalid per Cloudflare rules (would be silently coerced to 1).
      - If intended.proxied != live.proxied: proxy status itself drifted.
      - If intended.proxied is False and intended.ttl != live.ttl:
        real TTL drift, not the proxied-TTL quirk.
    """
    if live.get("proxied") is True and live.get("ttl") != 1:
        return "impossible state: live record is proxied but ttl is not 1 (stale read or cache)"

    if intended.get("proxied") is True and intended.get("ttl") not in (1, None):
        return "invalid config: proxied records are always coerced to ttl 1 by Cloudflare"

    if intended.get("proxied") != live.get("proxied"):
        return "proxy status drifted between intended config and live zone"

    if intended.get("proxied") is False and intended.get("ttl") != live.get("ttl"):
        return "real ttl drift on an unproxied record, not the proxied-ttl quirk"

    return None


def fetch_live_records(domain):
    """Return the live A/AAAA/CNAME records for domain from Cloudflare."""
    import requests

    headers = {"Authorization": f"Bearer {CLOUDFLARE_API_TOKEN}"}
    params = {"name": domain, "per_page": 100}
    r = requests.get(
        f"{CF_API}/zones/{CLOUDFLARE_ZONE_ID}/dns_records",
        headers=headers, params=params, timeout=30,
    )
    r.raise_for_status()
    return r.json()["result"]


def repair_record(record_id, domain, record_type, content, policy, desired_ttl=300):
    """Repair a mismatched record.

    policy: "accept_automatic" sets ttl to 1 and keeps proxied true.
            "unproxy" sets proxied to false and keeps desired_ttl.
    """
    import requests

    headers = {"Authorization": f"Bearer {CLOUDFLARE_API_TOKEN}"}
    if policy == "accept_automatic":
        body = {"type": record_type, "name": domain, "content": content, "proxied": True, "ttl": 1}
    elif policy == "unproxy":
        body = {"type": record_type, "name": domain, "content": content, "proxied": False, "ttl": desired_ttl}
    else:
        raise ValueError(f"unknown policy: {policy}")

    if DRY_RUN:
        log.info("[dry run] would PATCH record %s with %s", record_id, body)
        return

    r = requests.patch(
        f"{CF_API}/zones/{CLOUDFLARE_ZONE_ID}/dns_records/{record_id}",
        headers=headers, json=body, timeout=30,
    )
    r.raise_for_status()
    log.info("Repaired record %s with policy %s", record_id, policy)


def run(intended_by_name=None, policy=None):
    """intended_by_name: dict mapping record name -> {"ttl": int, "proxied": bool}."""
    if intended_by_name is None:
        intended_by_name = {f"app.{DNS_DOMAIN}": {"ttl": 300, "proxied": True}}
    if policy is None:
        policy = os.environ.get("REPAIR_POLICY", "accept_automatic")

    live_records = fetch_live_records(DNS_DOMAIN)
    mismatches = 0

    for rec in live_records:
        intended = intended_by_name.get(rec["name"])
        if intended is None:
            continue

        live = {"ttl": rec["ttl"], "proxied": rec["proxied"]}
        reason = diagnose_ttl_proxy_mismatch(intended, live)
        if reason is None:
            log.info("%s: consistent (ttl=%s, proxied=%s)", rec["name"], live["ttl"], live["proxied"])
            continue

        mismatches += 1
        log.warning("%s: %s", rec["name"], reason)

        if not (CLOUDFLARE_API_TOKEN and CLOUDFLARE_ZONE_ID):
            log.warning("No Cloudflare credentials set. Not repairing, only reporting.")
            continue

        repair_record(
            rec["id"], rec["name"], rec["type"], rec["content"],
            policy=policy, desired_ttl=intended.get("ttl") or 300,
        )

    log.info("Done. %d mismatch(es) found.", mismatches)


if __name__ == "__main__":
    run()

File: proxied-record-forces-ttl/node/proxied-record-forces-ttl.js
/**
 * Detect a proxied Cloudflare record whose intended config expects a
 * custom TTL, which Cloudflare will always silently coerce to 1
 * (Automatic), and optionally repair the zone by either accepting
 * ttl: 1 in policy, or unproxying the record to keep a custom TTL.
 *
 * Safe by default. Set DRY_RUN=false to let it write.
 *
 * Env vars:
 *   DNS_DOMAIN               the domain to check, e.g. "example.com"
 *   CLOUDFLARE_API_TOKEN     Cloudflare API token (only needed for repair)
 *   CLOUDFLARE_ZONE_ID       Cloudflare zone id (only needed for repair)
 *   DRY_RUN                  default "true"; set to "false" to actually write
 *   REPAIR_POLICY            "accept_automatic" or "unproxy", default "accept_automatic"
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "example.com";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const REPAIR_POLICY = process.env.REPAIR_POLICY || "accept_automatic";

const CF_API = "https://api.cloudflare.com/client/v4";

export function diagnoseTtlProxyMismatch(intended, live) {
  // Pure decision function. No I/O.
  //
  // intended, live: { ttl: number, proxied: boolean }
  //
  // Returns a mismatch reason string, or null if consistent.
  //   - If live.proxied is true and live.ttl !== 1: impossible/stale-cache state.
  //   - If intended.proxied is true and intended.ttl not in (1, null/undefined):
  //     config is invalid per Cloudflare rules (would be silently coerced to 1).
  //   - If intended.proxied !== live.proxied: proxy status itself drifted.
  //   - If intended.proxied is false and intended.ttl !== live.ttl:
  //     real TTL drift, not the proxied-TTL quirk.
  if (live.proxied === true && live.ttl !== 1) {
    return "impossible state: live record is proxied but ttl is not 1 (stale read or cache)";
  }

  const intendedTtl = intended.ttl;
  if (intended.proxied === true && intendedTtl !== 1 && intendedTtl != null) {
    return "invalid config: proxied records are always coerced to ttl 1 by Cloudflare";
  }

  if (intended.proxied !== live.proxied) {
    return "proxy status drifted between intended config and live zone";
  }

  if (intended.proxied === false && intendedTtl !== live.ttl) {
    return "real ttl drift on an unproxied record, not the proxied-ttl quirk";
  }

  return null;
}

async function fetchLiveRecords(domain) {
  const headers = { Authorization: `Bearer ${CLOUDFLARE_API_TOKEN}` };
  const params = new URLSearchParams({ name: domain, per_page: "100" });
  const res = await fetch(
    `${CF_API}/zones/${CLOUDFLARE_ZONE_ID}/dns_records?${params.toString()}`,
    { headers },
  );
  if (!res.ok) throw new Error(`Cloudflare list returned ${res.status}`);
  const body = await res.json();
  return body.result;
}

async function repairRecord(recordId, domain, recordType, content, policy, desiredTtl = 300) {
  const headers = {
    Authorization: `Bearer ${CLOUDFLARE_API_TOKEN}`,
    "Content-Type": "application/json",
  };

  let payload;
  if (policy === "accept_automatic") {
    payload = { type: recordType, name: domain, content, proxied: true, ttl: 1 };
  } else if (policy === "unproxy") {
    payload = { type: recordType, name: domain, content, proxied: false, ttl: desiredTtl };
  } else {
    throw new Error(`unknown policy: ${policy}`);
  }

  if (DRY_RUN) {
    console.log(`[dry run] would PATCH record ${recordId} with`, payload);
    return;
  }

  const res = await fetch(`${CF_API}/zones/${CLOUDFLARE_ZONE_ID}/dns_records/${recordId}`, {
    method: "PATCH",
    headers,
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error(`Cloudflare patch returned ${res.status}`);
  console.log(`Repaired record ${recordId} with policy ${policy}`);
}

export async function run(intendedByName, policy = REPAIR_POLICY) {
  // intendedByName: object mapping record name -> { ttl, proxied }
  if (!intendedByName) {
    intendedByName = { [`app.${DNS_DOMAIN}`]: { ttl: 300, proxied: true } };
  }

  const liveRecords = await fetchLiveRecords(DNS_DOMAIN);
  let mismatches = 0;

  for (const rec of liveRecords) {
    const intended = intendedByName[rec.name];
    if (!intended) continue;

    const live = { ttl: rec.ttl, proxied: rec.proxied };
    const reason = diagnoseTtlProxyMismatch(intended, live);
    if (!reason) {
      console.log(`${rec.name}: consistent (ttl=${live.ttl}, proxied=${live.proxied})`);
      continue;
    }

    mismatches++;
    console.warn(`${rec.name}: ${reason}`);

    if (!(CLOUDFLARE_API_TOKEN && CLOUDFLARE_ZONE_ID)) {
      console.warn("No Cloudflare credentials set. Not repairing, only reporting.");
      continue;
    }

    await repairRecord(
      rec.id, rec.name, rec.type, rec.content,
      policy, intended.ttl || 300,
    );
  }

  console.log(`Done. ${mismatches} mismatch(es) found.`);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}
